pipeline/authn/authenticator_oauth2_client_credentials.go

```diff
@@ -86,9 +95,92 @@ func (a *AuthenticatorOAuth2ClientCredentials) Config(config json.RawMessage) (*
 	timeout := time.Millisecond * duration
 	a.client = httpx.NewResilientClientLatencyToleranceConfigurable(nil, timeout, maxWait)
 
+	if c.Cache.TTL != "" {
+		cacheTTL, err := time.ParseDuration(c.Cache.TTL)
+		if err != nil {
+			return nil, err
+		}
+		a.cacheTTL = &cacheTTL
+	}
+
+	if a.tokenCache == nil {
+		cost := int64(c.Cache.MaxCost)
+		if cost == 0 {
+			cost = 100000000
+		}
+		a.logger.Debugf("Creating cache with max cost: %d", c.Cache.MaxCost)
+		cache, err := ristretto.NewCache(&ristretto.Config{
+			// This will hold about 1000 unique mutation responses.
+			NumCounters: 10000,
+			// Allocate a max
+			MaxCost: cost,
+			// This is a best-practice value.
+			BufferItems: 64,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		a.tokenCache = cache
+	}
+
 	return &c, nil
 }
 
+func clientCredentialsToKey(cc clientcredentials.Config) string {
+	return fmt.Sprintf("%s:%s", cc.ClientID, cc.ClientSecret)
+}
+
+func (a *AuthenticatorOAuth2ClientCredentials) tokenFromCache(config *AuthenticatorOAuth2Configuration, clientCredentials clientcredentials.Config) *oauth2.Token {
+	if !config.Cache.Enabled {
+		return nil
+	}
+
+	item, found := a.tokenCache.Get(clientCredentialsToKey(clientCredentials))
```

Comment:
This is very dangerous, the cache is not respecting the configuration at all. Imagine you have two introspection endpoints:

- `https://my-auth-server1.example.org`
- `https://my-auth-server2.example.org`

If you just look up by client credentials, once cached, the credential for auth-server1 will be valid at auth-server2 as well.